[validation-gen] fix unit test for validate/union

### What happened?

Ran unittest 
```
go test -v k8s.io/apimachinery/pkg/api/validate
```
has error
```
staging/src/k8s.io/apimachinery/pkg/api/validate/union_test.go:87:55: cannot use tc.fields (variable of type []any) as [][2]string value in argument to NewUnionMembership
staging/src/k8s.io/apimachinery/pkg/api/validate/union_test.go:87:70: not enough arguments in call to Union
        have (nil, *testUnion, *UnionMembership, []any)
        want (operation.Context, *field.Path, any, any, *UnionMembership, ...any)
```

Need to update the unittest here:
https://github.com/jpbetz/kubernetes/blob/a2e6b49acbf40c7017870ddc16a625759ef941b8/staging/src/k8s.io/apimachinery/pkg/api/validate/union_test.go#L87

### What did you expect to happen?

Successfully ran.

### How can we reproduce it (as minimally and precisely as possible)?

```
go test -v k8s.io/apimachinery/pkg/api/validate
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>


### Cloud provider

<details>

</details>


### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>


### Install tools

<details>

</details>


### Container runtime (CRI) and version (if applicable)

<details>

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

--- a/staging/src/k8s.io/apimachinery/pkg/api/validate/union.go
+++ b/staging/src/k8s.io/apimachinery/pkg/api/validate/union.go
@@ -86,7 +86,7 @@ func DiscriminatedUnion[T ~string](opCtx operation.Context, fldPath *field.Path,
 		isSpecified := rv.IsValid() && !rv.IsZero()
 		if isSpecified && !isDiscriminatedMember {
 			errs = append(errs, field.Invalid(fldPath.Child(member.fieldName), "",
-				fmt.Sprintf("may only be specified when `%s` is %q", union.discriminatorName, discriminatorValue)))
+				fmt.Sprintf("may only be specified when `%s` is %q", union.discriminatorName, member.memberName)))
 		} else if !isSpecified && isDiscriminatedMember {
 			errs = append(errs, field.Invalid(fldPath.Child(member.fieldName), "",
 				fmt.Sprintf("must be specified when `%s` is %q", union.discriminatorName, discriminatorValue)))

--- a/staging/src/k8s.io/apimachinery/pkg/api/validate/union_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/api/validate/union_test.go
@@ -20,6 +20,7 @@ import (
 	"reflect"
 	"testing"
 
+	"k8s.io/apimachinery/pkg/api/operation"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 )
 
@@ -29,62 +30,57 @@ type testMember struct{}
 func TestUnion(t *testing.T) {
 	testCases := []struct {
 		name        string
-		fields      []any
+		fields      [][2]string
 		fieldValues []any
 		expected    field.ErrorList
 	}{
 		{
 			name:        "valid pointers one of",
-			fields:      []any{"a", "b", "c", "d"},
+			fields:      [][2]string{{"a", "A"}, {"b", "B"}, {"c", "C"}, {"d", "D"}},
 			fieldValues: []any{nil, nil, nil, &testMember{}},
 			expected:    nil,
 		},
 		{
 			name:        "invalid pointers one of",
-			fields:      []any{"a", "b", "c", "d"},
+			fields:      [][2]string{{"a", "A"}, {"b", "B"}, {"c", "C"}, {"d", "D"}},
 			fieldValues: []any{nil, &testMember{}, nil, &testMember{}},
-			expected:    field.ErrorList{field.Invalid(nil, "", "must not specify \"b\" and \"d\", must specify one of: \"a\", \"b\", \"c\" or \"d\"")},
+			expected:    field.ErrorList{field.Invalid(nil, "{b, d}", "must specify exactly one of: `a`, `b`, `c`, `d`")},
 		},
 		{
 			name:        "valid string one of",
-			fields:      []any{"a", "b", "c", "d"},
+			fields:      [][2]string{{"a", "A"}, {"b", "B"}, {"c", "C"}, {"d", "D"}},
 			fieldValues: []any{"", "", "", "x"},
 			expected:    nil,
 		},
 		{
 			name:        "invalid string one of",
-			fields:      []any{"a", "b", "c", "d"},
+			fields:      [][2]string{{"a", "A"}, {"b", "B"}, {"c", "C"}, {"d", "D"}},
 			fieldValues: []any{"", "x", "", "x"},
-			expected:    field.ErrorList{field.Invalid(nil, "", "must not specify \"b\" and \"d\", must specify one of: \"a\", \"b\", \"c\" or \"d\"")},
+			expected:    field.ErrorList{field.Invalid(nil, "{b, d}", "must specify exactly one of: `a`, `b`, `c`, `d`")},
 		},
 		{
 			name:        "valid mixed type one of",
-			fields:      []any{"a", "b", "c", "d"},
+			fields:      [][2]string{{"a", "A"}, {"b", "B"}, {"c", "C"}, {"d", "D"}},
 			fieldValues: []any{0, "", nil, "x"},
 			expected:    nil,
 		},
 		{
 			name:        "invalid mixed type one of",
-			fields:      []any{"a", "b", "c", "d"},
+			fields:      [][2]string{{"a", "A"}, {"b", "B"}, {"c", "C"}, {"d", "D"}},
 			fieldValues: []any{0, "x", nil, &testMember{}},
-			expected:    field.ErrorList{field.Invalid(nil, "", "must not specify \"b\" and \"d\", must specify one of: \"a\", \"b\", \"c\" or \"d\"")},
+			expected:    field.ErrorList{field.Invalid(nil, "{b, d}", "must specify exactly one of: `a`, `b`, `c`, `d`")},
 		},
 		{
 			name:        "invalid no member set",
-			fields:      []any{"a", "b", "c", "d"},
+			fields:      [][2]string{{"a", "A"}, {"b", "B"}, {"c", "C"}, {"d", "D"}},
 			fieldValues: []any{nil, nil, nil, nil},
-			expected:    field.ErrorList{field.Invalid(nil, "", "must specify one of: \"a\", \"b\", \"c\" or \"d\"")},
-		},
-		{
-			name:        "valid multiple field for union member",
-			fields:      []any{[2]string{"a", "x"}, "b", "c", [2]string{"d", "x"}},
-			fieldValues: []any{1, nil, nil, 1},
+			expected:    field.ErrorList{field.Invalid(nil, "", "must specify exactly one of: `a`, `b`, `c`, `d`")},
 		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			got := Union(nil, &testUnion{}, NewUnionMembership(tc.fields...), tc.fieldValues...)
+			got := Union(operation.Context{Operation: operation.Update}, nil, nil, nil, NewUnionMembership(tc.fields...), tc.fieldValues...)
 			if !reflect.DeepEqual(got, tc.expected) {
 				t.Errorf("got %v want %v", got, tc.expected)
 			}
@@ -96,7 +92,7 @@ func TestDiscriminatedUnion(t *testing.T) {
 	testCases := []struct {
 		name               string
 		discriminatorField string
-		fields             []any
+		fields             [][2]string
 		discriminatorValue string
 		fieldValues        []any
 		expected           field.ErrorList
@@ -104,33 +100,26 @@ func TestDiscriminatedUnion(t *testing.T) {
 		{
 			name:               "valid discriminated union",
 			discriminatorField: "d",
-			fields:             []any{[2]string{"a", "A"}, "b", "c", [2]string{"d", "D"}},
+			fields:             [][2]string{{"a", "A"}, {"b", "B"}, {"c", "C"}, {"d", "D"}},
 			discriminatorValue: "A",
 			fieldValues:        []any{1, nil, nil, nil},
 		},
 		{
-			name:               "valid discriminated union, implicit union name",
-			discriminatorField: "type",
-			fields:             []any{[2]string{"a", "A"}, "b", "c", [2]string{"d", "D"}},
-			discriminatorValue: "B",
-			fieldValues:        []any{nil, 1, nil, nil},
-		},
-		{
 			name:               "invalid, discriminator not set to member that is specified",
 			discriminatorField: "type",
-			fields:             []any{[2]string{"a", "A"}, "b", "c", [2]string{"d", "D"}},
+			fields:             [][2]string{{"a", "A"}, {"b", "B"}, {"c", "C"}, {"d", "D"}},
 			discriminatorValue: "C",
 			fieldValues:        []any{nil, 1, nil, nil},
 			expected: field.ErrorList{
-				field.Invalid(field.NewPath("b"), "", "must not be specified when \"type\" is \"C\""),
-				field.Invalid(field.NewPath("c"), "", "must be specified when \"type\" is \"C\""),
+				field.Invalid(field.NewPath("b"), "", "may only be specified when `type` is \"B\""),
+				field.Invalid(field.NewPath("c"), "", "must be specified when `type` is \"C\""),
 			},
 		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			got := DiscriminatedUnion(nil, &testUnion{}, NewDiscriminatedUnionMembership(tc.discriminatorField, tc.fields...), tc.discriminatorValue, tc.fieldValues...)
+			got := DiscriminatedUnion(operation.Context{Operation: operation.Update}, nil, nil, nil, NewDiscriminatedUnionMembership(tc.discriminatorField, tc.fields...), tc.discriminatorValue, tc.fieldValues...)
 			if !reflect.DeepEqual(got, tc.expected) {
 				t.Errorf("got %v want %v", got.ToAggregate(), tc.expected.ToAggregate())
 			}

--- a/staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/discriminated_union/docs_test.go
+++ b/staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/discriminated_union/docs_test.go
@@ -29,7 +29,7 @@ func Test(t *testing.T) {
 	st.Value(&DU{D: DM2, M2: &M2{S: "x"}}).ExpectValid()
 
 	st.Value(&DU{D: DM2, M1: &M1{S: "x"}, M2: &M2{S: "x"}}).ExpectInvalid(
-		field.Invalid(field.NewPath("m1"), "", "may only be specified when `d` is \"M2\""),
+		field.Invalid(field.NewPath("m1"), "", "may only be specified when `d` is \"M1\""),
 	)
 
 	st.Value(&DU{D: DM1}).ExpectInvalid(

--- a/staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/discriminated_union_custom_members/docs_test.go
+++ b/staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/discriminated_union_custom_members/docs_test.go
@@ -29,7 +29,7 @@ func Test(t *testing.T) {
 	st.Value(&DU{D: DM2, M2: &M2{S: "x"}}).ExpectValid()
 
 	st.Value(&DU{D: DM2, M1: &M1{S: "x"}, M2: &M2{S: "x"}}).ExpectInvalid(
-		field.Invalid(field.NewPath("m1"), "", "may only be specified when `d` is \"CustomM2\""),
+		field.Invalid(field.NewPath("m1"), "", "may only be specified when `d` is \"CustomM1\""),
 	)
 
 	st.Value(&DU{D: DM1}).ExpectInvalid(

--- a/staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/multiple_discriminated_unions/docs_test.go
+++ b/staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/unions/multiple_discriminated_unions/docs_test.go
@@ -29,7 +29,7 @@ func Test(t *testing.T) {
 	st.Value(&DU{DU1: U1M2, U1M2: &M2{S: "x"}, DU2: U2M2, U2M2: &M2{S: "x"}}).ExpectValid()
 
 	st.Value(&DU{DU1: U1M2, U1M1: &M1{S: "x"}, U1M2: &M2{S: "x"}}).ExpectInvalid(
-		field.Invalid(field.NewPath("u1m1"), "", "may only be specified when `du1` is \"U1M2\""),
+		field.Invalid(field.NewPath("u1m1"), "", "may only be specified when `du1` is \"U1M1\""),
 	)
 
 	st.Value(&DU{DU1: U1M2}).ExpectInvalid(

